src/pal/pal_apple.h

```diff
@@ -1,92 +1,218 @@
       while (madvise(p, size, MADV_FREE_REUSABLE) == -1 && errno == EAGAIN)
         ;
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      // This must occur after `MADV_FREE_REUSABLE`.
+      //
+      // `mach_vm_protect` is observably slower in benchmarks.
+      mprotect(p, size, PROT_NONE);
+#  endif
     }
 
     /**
-     * same remark as above but we need to mark the page as REUSE
-     * first
+     * Notify platform that we will be using these pages.
      */
     template<ZeroMem zero_mem>
     static void notify_using(void* p, size_t size) noexcept
     {
       SNMALLOC_ASSERT(
-        is_aligned_block<PALBSD::page_size>(p, size) || (zero_mem == NoZero));
+        is_aligned_block<page_size>(p, size) || (zero_mem == NoZero));
+
+      if constexpr (zero_mem == YesZero)
+      {
+        void* r = mmap(
+          p,
+          size,
+          PROT_READ | PROT_WRITE,
+          MAP_PRIVATE | MAP_ANONYMOUS | MAP_FIXED,
+          anonymous_memory_fd,
+          0);
+
+        if (likely(r != MAP_FAILED))
+        {
+          return;
+        }
+      }
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      // Mark pages as writable for `madvise` below.
+      //
+      // `mach_vm_protect` is observably slower in benchmarks.
+      mprotect(p, size, PROT_READ | PROT_WRITE);
+#  endif
+
+      // `MADV_FREE_REUSE` can only be applied to writable pages,
+      // otherwise it's an error.
+      //
+      // `mach_vm_behavior_set` is observably slower in benchmarks.
+      //
+      // macOS 11 Big Sur may behave in an undocumented manner.
       while (madvise(p, size, MADV_FREE_REUSE) == -1 && errno == EAGAIN)
         ;
 
       if constexpr (zero_mem == YesZero)
-        zero<true>(p, size);
+      {
+        ::bzero(p, size);
+      }
     }
+
+    template<bool committed>
+    static void* reserve_aligned(size_t size) noexcept
+    {
+      SNMALLOC_ASSERT(bits::is_pow2(size));
+      SNMALLOC_ASSERT(size >= minimum_alloc_size);
+
+      // mask has least-significant bits set
+      mach_vm_offset_t mask = size - 1;
+
+      int flags = VM_FLAGS_ANYWHERE | default_mach_vm_map_flags;
+
+      // must be initialized to 0 or addr is interepreted as a lower-bound.
+      mach_vm_address_t addr = 0;
+
+#  ifdef USE_POSIX_COMMIT_CHECKS
+      vm_prot_t prot = committed ? VM_PROT_READ | VM_PROT_WRITE : VM_PROT_NONE;
+#  else
+      vm_prot_t prot = VM_PROT_READ | VM_PROT_WRITE;
 #  endif
+
+      kern_return_t kr = mach_vm_map(
+        mach_task_self(),
+        &addr,
+        size,
+        mask,
+        flags,
+        MEMORY_OBJECT_NULL,
+        0,
+        TRUE,
+        prot,
+        VM_PROT_READ | VM_PROT_WRITE,
+        VM_INHERIT_COPY);
+
+      if (unlikely(kr != KERN_SUCCESS))
+      {
+        error("Failed to allocate memory\n");
+      }
+
+      return reinterpret_cast<void*>(addr);
+    }
+
+    /**
+     * Source of Entropy
+     *
+     * Apple platforms do not have a public getentropy implementation, so use
```

Comment:
`arc4random_buf` can silently fail. `getentropy` and `CCGenerateRandomBytes` both return an error code.

https://opensource.apple.com/source/Libc/Libc-1439.40.11/gen/FreeBSD/arc4random.c.auto.html
```c
void
arc4random_buf(void *buf, size_t buf_size)
{
	arc4_init();
        // the error is ignored
	ccrng_generate(rng, buf_size, buf);
}
```

https://opensource.apple.com/source/CommonCrypto/CommonCrypto-60178.40.2/lib/CommonRandom.c.auto.html
```c
CCRNGStatus CCRandomGenerateBytes(void *bytes, size_t count)
{
    int err;
    struct ccrng_state *rng;

    if (0 == count) {
        return kCCSuccess;
    }

    if (NULL == bytes) {
        return kCCParamError;
    }

    rng = ccDRBGGetRngState();
    // the error is captured
    err = ccrng_generate(rng, count, bytes);
    if (err == CCERR_OK) {
        return kCCSuccess;
    }

    return kCCRNGFailure;
}
```

I found an app that links against `arc4random` on the App Store.

Reply:
I'll correct the comment so that it is more clear and concise.

Also, `arc4random` is not seeded with `genentropy`:

https://opensource.apple.com/source/Libc/Libc-1439.40.11/gen/FreeBSD/arc4random.c.auto.html
```c
static struct ccrng_state *rng;

static void
arc4_init(void)
{
	int err;

	if (rng != NULL) return;

	rng = ccrng(&err);
	if (rng == NULL) {
#if OS_CRASH_ENABLE_EXPERIMENTAL_LIBTRACE
		os_crash("arc4random: unable to get ccrng() handle (%d)", err);
#else
		os_crash("arc4random: unable to get ccrng() handle");
#endif
	}
}

uint32_t
arc4random(void)
{
	uint32_t rand;

	arc4random_buf(&rand, sizeof(rand));

	return rand;
}

void
arc4random_buf(void *buf, size_t buf_size)
{
	arc4_init();
	ccrng_generate(rng, buf_size, buf);
}
```

Reply:
The dependency is CommonCrypto. CoreCrypto is private. Yes, `ccrng_generate` is a private function.

CommonCrypto (`/usr/lib/system/libcommonCrypto.dylib`) is implicitly linked by `dyld` (macOS linker.) We have no say in the matter as library developers. If there's any "library initialization" that is dependent on malloc in CommonCrypto, I'd argue that the behavior is undefined and is intrinsic to macOS in production.

I'd argue that Apple makes no guarantees about future behavior of standard system APIs, as we saw with the macOS 11.2 release (a minor release!) that broke mprotect (W^X) and many projects experienced issues.

The descriptions of ccrng and ccrng_generate:

https://github.com/apple/darwin-xnu/blob/main/EXTERNAL_HEADERS/corecrypto/ccrng.h
```c
/*!
 @function   ccrng
 @abstract   Initializes an AES-CTR mode cryptographic random number generator and returns the statically-allocated rng object.
             Getting a pointer to a ccrng has never been simpler!
             Call this function, get an rng object and then pass the object to ccrng_generate() to generate randoms.
             ccrng() may be called more than once. It returns pointer to the same object on all calls.
 @result  a cryptographically secure random number generator or NULL if fails
 @discussion
 - It is significantly faster than using the system /dev/random
 - FIPS Compliant: NIST SP800-90A + FIPS 140-2
 - Seeded from the system entropy.
 - Provides at least 128bit security if the system provide 2bit of entropy / byte.
 - Entropy accumulation
 - Backtracing resistance
 - Prediction break with frequent (asynchronous) reseed
 */
```
```c
/*!
 @function   ccrng_generate
 @abstract   Generate `outlen` bytes of output, stored in `out`, using ccrng_state `rng`.
 @param rng  `struct ccrng_state` representing the state of the RNG.
 @param outlen  Amount of random bytes to generate.
 @param out  Pointer to memory where random bytes are stored, of size at least `outlen`.
 @result 0 on success and nonzero on failure.
 */
```

ccrng "Initializes an AES-CTR mode cryptographic random number generator and returns the **statically-allocated rng object** [emphasis mine]."

Standards-compliant entities/organizations deploy to macOS and iOS using this API. See, https://www.niap-ccevs.org/Documents_and_Guidance/view_td.cfm?TD=0510

Reply:
`ccrng` is closed source. `ccrng_generate` calls the function pointer in `ccrng_state`.

Reply:
> If there is an error, we are killing the process.

I would much sooner kill the process.  We build different versions of the library with and without the mitigations.  If it fails with mitigations, then the user can try the version without mitigations.  Silent failure seems like a problem for security auditing.